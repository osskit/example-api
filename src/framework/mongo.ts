import type { Db } from 'mongodb';
import { MongoClient, ReadPreference } from 'mongodb';
import { logger } from './logger.js';

import { mongoDb, mongoUri } from './environment.js';

let connection: {
  db: Db;
  client: MongoClient;
};

export const defaultProjection = { _id: 0 } as const;

const connect = ({ uri, dbName }: { uri: string; dbName: string }) =>
  MongoClient.connect(`${uri}/${dbName}`, {
    readPreference: ReadPreference.PRIMARY,
    retryWrites: true,
    w: 'majority',
  });

export const init = async () => {
  try {
    const client = await connect({ uri: mongoUri, dbName: mongoDb });

    connection = {
      client,
      db: client.db(mongoDb),
    };

    logger.info('connected to mongo');
  } catch (error) {
    throw new Error(`couldn't connect to mongo ${error}`);
  }
};

export const getCollection = <T>(collectionName: string) => connection.db.collection<T>(collectionName);

export const ping = () => connection.db.command({ ping: 1 });

export const close = () => connection.client.close();
