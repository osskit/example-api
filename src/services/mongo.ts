import { createMonitor } from '@osskit/monitor';
import { ReturnDocument } from 'mongodb';
import { getCollection } from '../framework/mongo.js';
import type { Pokemon, UUID } from '../schemas/models.js';
import type { Meta } from '../schemas/api.js';

const monitor = createMonitor({ scope: 'mongo' });

const getPokemonCollection = () => getCollection<Pokemon>('pokemon');

export const getPokemon = (id: UUID, meta: Meta): Promise<Pokemon | null> =>
  monitor('getPokemon', () => getPokemonCollection().findOne({ id }), { context: { id, meta } });

export const upsertPokemon = (pokemon: Pokemon, meta: Meta): Promise<Pokemon> =>
  monitor(
    'upsertPokemon',
    async () => {
      const result = await getPokemonCollection().findOneAndUpdate(
        pokemon,
        {
          $set: pokemon,
        },
        {
          upsert: true,
          returnDocument: ReturnDocument.AFTER,
        },
      );

      if (!result.ok) {
        throw new Error('Pokemon could not created');
      }

      return result.value!;
    },
    {
      context: {
        pokemon,
        meta,
      },
    },
  );
