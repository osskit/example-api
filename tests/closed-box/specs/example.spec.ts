import fetchApi from 'node-fetch';
import { Configuration, ExamplesApi } from '../../../packages/client/index.js';
import { clear as clearMongo, close as closeMongo, init as initMongo, insertExamples } from '../mongo.js';

const testExample = {
  id: '6740239e-a263-4933-9c8e-b397ba473fa5',
  name: 'name',
  type: 'a',
};

const exampleApi = new ExamplesApi(
  new Configuration({
    basePath: 'http://localhost:3000/v1',
    fetchApi: fetchApi as any,
    headers: {
      'x-api-client': 'test',
      'x-api-client-version': '1',
    },
  }),
);

describe('main', () => {
  beforeAll(async () => {
    await initMongo();
  });

  beforeEach(async () => {
    await clearMongo();
  });

  afterAll(async () => {
    await closeMongo();
  });

  it('get example', async () => {
    await insertExamples([testExample]);
    const example = await exampleApi.getExample(testExample.id);

    expect(example).toMatchSnapshot();
  });

  it('create example', async () => {
    const example = await exampleApi.createExample({ type: testExample.type, name: testExample.name });

    expect(example).toMatchSnapshot({ id: expect.any(String) });
  });
});
