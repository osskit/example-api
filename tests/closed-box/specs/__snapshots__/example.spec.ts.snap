// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`main create example 1`] = `
Object {
  "id": Any<String>,
  "name": "name",
  "type": "a",
}
`;

exports[`main get example 1`] = `
Object {
  "id": "6740239e-a263-4933-9c8e-b397ba473fa5",
  "name": "name",
  "type": "a",
}
`;
